client/bgp_instance.go

```diff
@@ -0,0 +1,127 @@
+
+import (
+	"fmt"
+	"log"
+	"strings"
+)
+
+// BgpInstance Mikrotik resource
+type BgpInstance struct {
+	ID                       string `mikrotik:".id"`
+	Name                     string `mikrotik:"name"`
+	As                       int    `mikrotik:"as"`
+	ClientToClientReflection bool   `mikrotik:"client-to-client-reflection"`
+	Comment                  string `mikrotik:"comment"`
+	ConfederationPeers       string `mikrotik:"confederation-peers"`
+	Disabled                 bool   `mikrotik:"disabled"`
+	IgnoreAsPathLen          bool   `mikrotik:"ignore-as-path-len"`
+	OutFilter                string `mikrotik:"out-filter"`
+	RedistributeConnected    bool   `mikrotik:"redistribute-connected"`
+	RedistributeOspf         bool   `mikrotik:"redistribute-ospf"`
+	RedistributeOtherBgp     bool   `mikrotik:"redistribute-other-bgp"`
+	RedistributeRip          bool   `mikrotik:"redistribute-rip"`
+	RedistributeStatic       bool   `mikrotik:"redistribute-static"`
+	RouterID                 string `mikrotik:"router-id"`
+	RoutingTable             string `mikrotik:"routing-table"`
+	ClusterID                string `mikrotik:"cluster-id"`
+	Confederation            int    `mikrotik:"confederation"`
+}
+
+// AddBgpInstance Mikrotik resource
+func (client Mikrotik) AddBgpInstance(b *BgpInstance) (*BgpInstance, error) {
+	c, err := client.getMikrotikClient()
+
+	if err != nil {
+		return nil, err
+	}
+
+	attributes := Marshal(b)
+	cmd := strings.Split(fmt.Sprintf("/routing/bgp/instance/add %s", attributes), " ")
+
+	log.Printf("[INFO] Running the mikrotik command: `%s`", cmd)
+	r, err := c.RunArgs(cmd)
+	log.Printf("[DEBUG] /routing/bgp/instance/add returned %v", r)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return client.FindBgpInstance(b.Name)
+}
+
+// FindBgpInstance Mikrotik resource
+func (client Mikrotik) FindBgpInstance(name string) (*BgpInstance, error) {
+	c, err := client.getMikrotikClient()
+	if err != nil {
+		return nil, err
+	}
+
+	cmd := strings.Split(fmt.Sprintf("/routing/bgp/instance/print ?name=%s", name), " ")
+	log.Printf("[INFO] Running the mikrotik command: `%s`", cmd)
+	r, err := c.RunArgs(cmd)
+
+	if err != nil {
+		return nil, err
+	}
+
+	log.Printf("[DEBUG] Find bgp instance: `%v`", cmd)
+
+	bgpInstance := BgpInstance{}
+
+	err = Unmarshal(*r, &bgpInstance)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if bgpInstance.Name == "" {
+		return nil, NewNotFound(fmt.Sprintf("bgp instance `%s` not found", name))
+	}
+
+	return &bgpInstance, nil
+}
+
+// UpdateBgpInstance Mikrotik resource
+func (client Mikrotik) UpdateBgpInstance(b *BgpInstance) (*BgpInstance, error) {
+	c, err := client.getMikrotikClient()
+
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = client.FindBgpInstance(b.Name)
+
+	if err != nil {
+		return b, err
+	}
+	attributes := Marshal(b)
+	cmd := strings.Split(fmt.Sprintf("/routing/bgp/instance/set %s", attributes), " ")
+
+	log.Printf("[INFO] Running the mikrotik command: `%s`", cmd)
+	_, err = c.RunArgs(cmd)
```

Comment:
I've always thought this command construction was very ugly. I think we could have these structs implement an interface to make it better.

```go
interface MikrotikCommand {
  func GetCommandNamespace() string // this will return "/routing/bgp/instance"
  func GetFindAttribute() (string, string) // this will return "name", "bgp-resource-name" which could be composed to "/routing/bgp/instance/print ?name=%s"
  func GetRemoveAttribute() (string, string) // hopefully you see the point (would help build the "remove" command
}
```

I need to put more thought into this but have you considered ways to improve this command construction?

Reply:
I truthfully hadn't thought about it since I had already spent time introducing the new `Marshal` function in trying to add support for the BGP Instance resource. I'll give it some thought as well, but I do agree with and like the general approach you suggested above.